### PR TITLE
1832 add tutorial for temporal encoding

### PR DESCRIPTION
**Pull request by p-offtermatt**

### CODE REVIEW

**Review by thpani — COMMENTED** [review]
This is a great writeup!

I think it's all understandable, though I don't know what a person without prior knowledge about model checking algorithms would make of it. Maybe someone who can read LTL but doesn't know the CTL/LTL model checking algorithms should review this as well – @konnov WDYT?

Two major issues that I'd like to discuss:
1. The TLA+ model differs a bit from the prose specification / a real-life traffic light. I think we should at least account for that.
2. The explanation of `_unroll` variables needs a bit of work, I think. AFAIR these are introduced to encode that `[]f` holds in all states of the loop (as opposed to a single state for the subformula auxiliary variable), right? From the description here it's not obvious if this is only relevant for `[]` subformulae.

Otherwise, just some suggestions for wording etc. in the comments.

Also, are you sure you don't want to merge this directly into unstable?

File: docs/src/tutorials/temporal-properties.md
Comment (p-offtermatt) [reply docs/src/tutorials/temporal-properties.md:365]: I added more explanation.

Comment (thpani) [reply docs/src/tutorials/temporal-properties.md:365]: I'm afraid I still don't understand.  This is probably obvious if one knows the entire construction, which I don't.
To clarify, is `☐(requestedGreen ⇒ ♢isGreen)` in the first state a prophecy variable, or is it constrained st we get a model value for it after finding a counter-example / exhaustively exploring the search space?

Comment (p-offtermatt) [reply docs/src/tutorials/temporal-properties.md:360]: It is a "prohecy variable". If the variable is false, then
in this trace, the formula will not hold. However, only a trace where the prohecy variable says false, and the model checker finds a valid loop (satisfying some constraints wrt the model and aux. variables) is a problem, so the prophecy variable being false does not yet constitute a bad trace. It being false and then later being able to find a valid loop is a counterexample.

Comment (thpani) [reply docs/src/tutorials/temporal-properties.md:360]: I see, thanks for the explanation!
It seems to me that there is no way of motivating the `_init` variable without knowing the construction. Which would be useful for someone to understand the construction, but this tutorial is aimed at users checking temporal properties. Maybe we should just say that it is an artifact of the model checking algorithm?

Comment (thpani) [reply docs/src/tutorials/temporal-properties.md:360]: We could give an intuition, e.g., in all states `_init` encodes whether the property holds in the first state, without giving details about why this is necessary.

**Comment by p-offtermatt** [comment]
Thanks for the great feedback Thomas!
I've incorporated the suggestions and rewritten some things, take a look if you want.
As for merging into unstable: This has a "soft dependency" on the counterexample writer, since otherwise the produced counterexamples look nothing like in the tutorial. So I'll merge this after the counterexample writer is in.

File: docs/src/tutorials/temporal-properties.md
Comment (thpani) [reply docs/src/tutorials/temporal-properties.md:360]: I think this is still an issue.

WDYT about my suggestion above?
> Maybe we should just say that it is an artifact of the model checking algorithm?
>  We could give an intuition, e.g., in all states `_init` encodes whether the property holds in the first state, without giving details about why this is necessary.

Comment (p-offtermatt) [reply docs/src/tutorials/temporal-properties.md]: My bad, one commit got lost on the way. I've added a paragraph.

File: docs/src/tutorials/temporal-properties.md
Comment (thpani) [reply docs/src/tutorials/temporal-properties.md:262]: This is not resolved yet.

File: docs/src/tutorials/temporal-properties.md
Comment (thpani) [reply docs/src/tutorials/temporal-properties.md]: looks good, thx!

**Comment by thpani** [comment]
@p-offtermatt I think there's still some open issues; I've marked the respective comments unresolved in my original review. Can you please check and address them inline?

File: docs/src/tutorials/temporal-properties.md
Comment (thpani) [reply docs/src/tutorials/temporal-properties.md:295]: I don't understand what state "we won't see the state of the loop again".

Also, if something is done by an algorithm, I prefer to say it that we instead of using a generic "we".

**Comment by p-offtermatt** [comment]
@thpani I just noticed why I thought commits were missing - I pushed to the wrong branch... 😔
Sorry about that - I think my workflow of resolving discussions when I address locally isn't working so well.
My bad, I'll cherry pick the commits in later.

**Comment by thpani** [comment]
@p-offtermatt Oh I see, no worries!
Git can be a tricky beast :)

File: docs/src/tutorials/temporal-properties.md
Comment (p-offtermatt) [reply docs/src/tutorials/temporal-properties.md]: Rewrote this a bit and replaced "we" where appropriate

File: docs/src/tutorials/temporal-properties.md
Comment (p-offtermatt) [reply docs/src/tutorials/temporal-properties.md]: I added how this works for `<>` operators

File: docs/src/tutorials/temporal-properties.md
Comment (p-offtermatt) [reply docs/src/tutorials/temporal-properties.md:58]: Added this below

File: docs/src/tutorials/temporal-properties.md
Comment (p-offtermatt) [reply docs/src/tutorials/temporal-properties.md:256]: Should be there now

File: docs/src/tutorials/temporal-properties.md
Comment (p-offtermatt) [inline docs/src/tutorials/temporal-properties.md:198]:
```suggestion
we store the valuation of all variables in an auxiliary copy whose name is prefixed by `__loop_`.
```

```suggestion
we store the valuation of all variables in an auxiliary copy whose name is prefixed by `__loop_`.
Before the first state of the loop, the values of the `__loop_` variables is unspecified.
```

File: docs/src/tutorials/TrafficLight.tla
Comment (p-offtermatt) [reply docs/src/tutorials/TrafficLight.tla:30]: I explicitly talk about the simplifications in the text now.

File: docs/src/tutorials/temporal-properties.md
Comment (konnov) [inline docs/src/tutorials/temporal-properties.md:175]: The notion of a trace appears out of nowhere. We usually use the word "computation" and the canonical TLA+ notion is that of a behavior. I think you are right in that we have to talk about traces here, as we decorate computation states with additional predicates. It would be good to have the notion of a trace quickly introduced in one paragraph.

File: docs/src/tutorials/temporal-properties.md
Comment (konnov) [inline docs/src/tutorials/temporal-properties.md:195]: Let's use the word "lasso" for executions with a loop, to have the stable vocabulary.

File: docs/src/tutorials/temporal-properties.md
Comment (konnov) [inline docs/src/tutorials/temporal-properties.md:198]:
```suggestion
we store the valuation of all variables in a shadow copy whose name is prefixed by `__loop_`.
```

File: docs/src/tutorials/temporal-properties.md
Comment (konnov) [inline docs/src/tutorials/temporal-properties.md:199]:
```suggestion
Before the first state of the loop, the values of the `__loop_` variables may carry arbitrary values.
```

They are specified in the counterexample.

File: docs/src/tutorials/temporal-properties.md
Comment (konnov) [inline docs/src/tutorials/temporal-properties.md:9]: I wanted to give a link to a video lecture by Leslie Lamport on that topic, but it looks like there is no lecture where he is presenting `<>`, or maybe I have not found it. Maybe give a reference to "Specifying Systems"?

File: docs/src/tutorials/temporal-properties.md
Comment (konnov) [inline docs/src/tutorials/temporal-properties.md:4]: Add yourself as an author.

File: docs/src/tutorials/temporal-properties.md
Comment (konnov) [inline docs/src/tutorials/temporal-properties.md:293]:
```suggestion
formula `[](requestedGreen => <>isGreen)` holds. Since we are looking at a counterexample to this formula, it should not be
surprising that the formula does not hold in the initial state of the counterexample.
```

**Review by konnov — APPROVED** [review]
Great tutorial. I have left a few comments. IMO, we can merge it and then think about further improvements. For instance, we could present the counterexamples as a table (since it is the tableaux encoding) or draw a trace with the values of the predicates below the states. But let's keep it for follow-up PRs.

File: docs/src/tutorials/temporal-properties.md
Comment (thpani) [reply docs/src/tutorials/temporal-properties.md:9]: There is a brief explanation of `<>` in video 9a.
But maybe Hillel's page on temporal operators is more informative? https://learntla.com/temporal-logic/operators/

**Review by thpani — APPROVED** [review]
LGTM, modulo Igor's comments.

> For instance, we could present the counterexamples as a table (since it is the tableaux encoding) or draw a trace with the values of the predicates below the states. But let's keep it for follow-up PRs.

Good ideas! @p-offtermatt, can you make sure we track them in an issue?

File: docs/src/tutorials/temporal-properties.md
Comment (p-offtermatt) [reply docs/src/tutorials/temporal-properties.md:9]: Added a link to Hilel's page

File: docs/src/tutorials/temporal-properties.md
Comment (p-offtermatt) [reply docs/src/tutorials/temporal-properties.md:179]: The nomenclature is from https://apalache.informal.systems/docs//tutorials/symbmc.html?highlight=trace#symbolic-traces, so I'll just link that

**Comment by p-offtermatt** [comment]
I've addressed Igor's comments and added an issue for general improvements on the tutorial:
 #1877

**Pull request merged**
